src/build/common/FileGenerator.js

```diff
@@ -0,0 +1,177 @@
+var mkdirp = require('mkdirp');
+
+var crypto = require('crypto');
+
+// Only rewrite files if needed
+
+var hashString = function(str) {
+  var md5sum = crypto.createHash('md5');
+  md5sum.update(str);
+  return md5sum.digest('hex');
+};
+
+var runGenerator = function(opts) {
+  var doWrite = function(err, src) {
+    mkdirp(path.dirname(opts.outputPath), function(err) {
+      if (err) { opts.cb(err); return; }
+
+      fs.writeFile(opts.outputPath, src, function(err) {
+        if (err) { opts.cb(err); return; }
+
+        // For dynamic calls the hash should be that of the input string, not the output file
+        if (opts.useInputHash) {
+          fs.writeFile(opts.outputHashPath, opts.inputHash, function(err) {
+            if (err) { opts.cb(err); return; }
+            opts.cb(null, src);
+          });
+        } else {
+          opts.cb(null, src);
+        }
+      });
+
+    });
+  };
+
+  var useOldOutput = function() {
+    // Old one is still good, just read it
+    fs.readFile(opts.outputPath, 'utf8', function(err, src) {
+      if (err) { opts.cb(err); return; }
+
+      opts.cb(null, src);
+    });
+  }
+
+  var checkModifiedTimes = function() {
+    fs.stat(opts.sourcePath, function(err, srcStat) {
+      if (err) { opts.cb(err); return; }
+
+      fs.stat(opts.outputPath, function(err, existingStat) {
+        if (err) { opts.cb(err); return; }
+
+        if (existingStat.mtime > srcStat.mtime) {
+          useOldOutput();
+          return;
+        }
+
+        opts.generateFn(doWrite);
+      });
+    });
+  };
+
+  var checkInputHash = function() {
+    // Cheating: add to opts so we have it inside of doWrite
+    opts.useInputHash = true;
+    opts.outputHashPath = opts.outputPath + '.hash';
+    opts.inputHash = hashString(opts.sourceContents);
+
+    // Get the output hash
+    fs.exists(opts.outputHashPath, function(exists) {
+      if (exists) {
+
+        // Check the hashes
+        fs.readFile(opts.outputHashPath, 'utf-8', function(err, outputHash) {
+          if (err) { opts.cb(err); return; }
+
+          if (opts.inputHash === outputHash) {
+            useOldOutput();
+            return;
+          }
+
+          opts.generateFn(doWrite);
+        });
+
+      } else {
+        opts.generateFn(doWrite);
+      }
+    });
+  };
+
+  // Check if the output exists
+  fs.exists(opts.outputPath, function(exists) {
+    if (exists) {
+      if (opts.sourcePath !== undefined) {
+        // It does, check the modified times
+        checkModifiedTimes();
+      } else if (opts.sourceContents !== undefined) {
+        checkInputHash();
+      } else {
+        throw new Error('unknown input');
+      }
+    } else {
+      opts.generateFn(doWrite);
+    }
+  });
+};
+
+
+/** take an input file (source), and generate the output with generateFn. finally
+run cb with cb(err, src). Will only run generateFn if the output file is older than
+the source file */
+module.exports = function(source, output, generateFn, cb) {
```

Comment:
files that export a function should be lower-cased (upper-cased files export classes)